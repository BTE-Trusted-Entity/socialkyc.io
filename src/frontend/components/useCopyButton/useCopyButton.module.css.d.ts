export const __esModule: true;
export const copy: string;
export const copied: string;
