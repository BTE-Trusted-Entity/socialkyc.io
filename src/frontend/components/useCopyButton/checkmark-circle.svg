<svg width="12" height="12" xmlns="http://www.w3.org/2000/svg"><g fill="#5AB9AA"><path d="M6 0C2.7 0 0 2.7 0 6s2.7 6 6 6 6-2.7 6-6-2.7-6-6-6zm0 11.006A5.017 5.017 0 0 1 .994 6 5.017 5.017 0 0 1 6 .994 5.017 5.017 0 0 1 11.006 6 5.017 5.017 0 0 1 6 11.006z"/><path d="m7.65 4.144-2.306 2.25-1.05-.994c-.206-.206-.506-.206-.694 0-.188.206-.206.506 0 .694l1.256 1.2c.15.15.3.206.45.206.15 0 .356-.056.45-.206l2.55-2.456c.207-.207.207-.507 0-.694-.15-.188-.45-.188-.656 0z"/></g></svg>
