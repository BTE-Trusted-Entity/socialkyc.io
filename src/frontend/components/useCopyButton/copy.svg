<svg width="18" height="22" xmlns="http://www.w3.org/2000/svg"><g fill="#E43366" stroke="#FFF" fill-rule="evenodd"><rect x="3.5" y=".5" width="14" height="17" rx="3"/><rect x=".5" y="4.5" width="14" height="17" rx="3"/></g></svg>