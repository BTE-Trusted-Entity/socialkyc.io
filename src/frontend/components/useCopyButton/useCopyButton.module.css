.copy {
  border: 0;
  padding: 1.75rem 0 0;
  display: inline-flex;
  cursor: pointer;
  background: url('copy.svg') no-repeat center top;
  width: 2rem;
  height: 3rem;
  right: 0.25rem;
  font-size: 0.875rem;
  color: var(--color-white);
  position: relative;
}

.copied {
  composes: copy;
}

.copied::after {
  content: 'copied to clipboard';
  width: 9em;
  position: absolute;
  top: 2.75em;
  left: -3.5em;
  display: inline-block;
  padding: 0.5em 0.75em;
  background: var(--color-magenta);
  border: 1px solid var(--color-light);
  border-radius: 0.75em;
  box-shadow: -2px -2px 4px 0 rgb(var(--color-white-rgb) / 30%);
}
