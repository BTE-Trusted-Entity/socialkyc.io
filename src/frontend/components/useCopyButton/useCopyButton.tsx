import { MouseEventHandler, RefObject, useCallback, useState } from 'react';

import * as styles from './useCopyButton.module.css';

interface UseCopyButton {
  supported: boolean;
  handleCopyClick: MouseEventHandler;
  className: string;
  title: string;
  justCopied: boolean;
}

export function useCopyButton(
  inputRef: RefObject<HTMLInputElement | HTMLTextAreaElement>,
): UseCopyButton {
  const writeTextSupported =
    navigator.clipboard && 'writeText' in navigator.clipboard;
  const supported =
    writeTextSupported || document.queryCommandSupported('copy');

  const [justCopied, setJustCopied] = useState(false);
  const className = justCopied ? styles.copied : styles.copy;

  const handleCopyClick = useCallback(async () => {
    const input = inputRef.current;
    if (!input || !supported) {
      return;
    }

    if (writeTextSupported) {
      await navigator.clipboard.writeText(input.value);
    } else {
      input.select();
      document.execCommand('copy');
    }

    setJustCopied(true);
    setTimeout(() => {
      setJustCopied(false);
    }, 500);
  }, [inputRef, supported, writeTextSupported]);

  return {
    supported,
    handleCopyClick,
    className,
    title: 'Copy',
    justCopied,
  };
}
