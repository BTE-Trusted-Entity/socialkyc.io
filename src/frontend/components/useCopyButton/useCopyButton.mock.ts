import { jest } from '@jest/globals';

import { useCopyButton } from './useCopyButton';

jest.mock('./useCopyButton');

jest.mocked(useCopyButton).mockImplementation(() => ({
  supported: true,
  handleCopyClick: jest.fn(),
  className: 'copy',
  title: 'copy to clipboard',
  justCopied: false,
}));
