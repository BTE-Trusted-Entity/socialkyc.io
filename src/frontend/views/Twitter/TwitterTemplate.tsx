import {
  FormEvent,
  FormEventHandler,
  Fragment,
  MouseEventHandler,
  useCallback,
  useRef,
  useState,
} from 'react';
import { Prompt } from 'react-router-dom';
import cx from 'classnames';

import * as flowStyles from '../../components/CredentialFlow/CredentialFlow.module.css';
import * as styles from './Twitter.module.css';

import { Spinner } from '../../components/Spinner/Spinner';
import { Explainer } from '../../components/Explainer/Explainer';
import { ExpiryDate } from '../../components/ExpiryDate/ExpiryDate';
import { SigningErrorClosed } from '../../components/SigningErrorClosed/SigningErrorClosed';
import { DetailedMessage } from '../../components/DetailedMessage/DetailedMessage';
import { AttestationProcessAnchoring } from '../../components/AttestationProcessAnchoring/AttestationProcessAnchoring';
import { AttestationProcessReady } from '../../components/AttestationProcessReady/AttestationProcessReady';
import { AttestationErrorUnknown } from '../../components/AttestationErrorUnknown/AttestationErrorUnknown';
import { LinkBack } from '../../components/LinkBack/LinkBack';
import { usePreventNavigation } from '../../utilities/usePreventNavigation';
import { useCopyButton } from '../../components/useCopyButton/useCopyButton';

export type AttestationStatus =
  | 'none'
  | 'requested'
  | 'confirming'
  | 'unconfirmed'
  | 'attesting'
  | 'ready'
  | 'error';

export type FlowError = 'closed' | 'unknown';

interface Props {
  status: AttestationStatus;
  processing: boolean;
  handleSubmit: FormEventHandler;
  handleBackup: MouseEventHandler;
  handleTryAgainClick: MouseEventHandler;
  setInputError: (error?: string) => void;
  inputError?: string;
  flowError?: FlowError;
  secret?: string;
}

export function TwitterTemplate({
  status,
  processing,
  handleSubmit,
  handleBackup,
  handleTryAgainClick,
  setInputError,
  inputError,
  flowError,
  secret,
}: Props): JSX.Element {
  const [twitterHandle, setTwitterHandle] = useState('');

  const handleInput = useCallback(
    (event: FormEvent<HTMLInputElement>) => {
      setTwitterHandle((event.target as HTMLInputElement).value);
      setInputError(undefined);
    },
    [setInputError],
  );

  const handleBlur = useCallback(() => {
    const trimmed = twitterHandle.trim();
    if (trimmed !== twitterHandle) {
      setTwitterHandle(trimmed);
    }
  }, [twitterHandle]);

  const preventNavigation = usePreventNavigation(
    processing || ['confirming', 'attesting'].includes(status),
  );

  const messageRef = useRef<HTMLTextAreaElement>(null);
  const copy = useCopyButton(messageRef);

  return (
    <section
      className={cx(flowStyles.container, {
        [flowStyles.processing]: processing,
      })}
      aria-busy={processing}
    >
      {processing && <Spinner />}

      <h1 className={styles.heading}>Twitter Account Attestation</h1>

      <Prompt
        when={preventNavigation}
        message="The Twitter attestation process has already started. Are you sure you want to leave?"
      />

      <Explainer>
        After entering your Twitter handle, please choose an identity in your
        wallet to associate with your Twitter credential. We will prompt you to
        Tweet so that we can attest your credential.
      </Explainer>

      <section>
        {(status === 'none' || status === 'requested') && (
          <form onSubmit={handleSubmit}>
            <label>
              Your Twitter handle
              <span
                className={styles.twitterInputContainer}
                data-invalid={inputError}
              >
                <input
                  className={styles.twitterInput}
                  onInput={handleInput}
                  onBlur={handleBlur}
                  value={twitterHandle}
                  type="text"
                  name="twitterHandle"
                  required
                  maxLength={100}
                />
              </span>
            </label>

            <output className={styles.inputError} hidden={!inputError}>
              {inputError}
            </output>

            <p>
              Validity: one year (<ExpiryDate />)
            </p>

            {flowError === 'closed' && <SigningErrorClosed />}

            <button
              type="submit"
              className={flowStyles.ctaButton}
              disabled={!twitterHandle}
            >
              Continue in wallet
            </button>
          </form>
        )}

        {status === 'confirming' && (
          <Fragment>
            <DetailedMessage
              icon="spinner"
              heading="Attestation process:"
              message="Starting"
              details="Your credential will be attested as soon as you Tweet the text below."
            />
            <div>
              <label htmlFor="tweet">Please tweet this message:</label>
              <p className={styles.tweetContainer}>
                <textarea
                  className={styles.tweetInput}
                  id="tweet"
                  ref={messageRef}
                  value={`I use SocialKYC to represent my internet identity. @social_kyc_tech ${secret}`}
                  readOnly
                />
                {copy.supported && (
                  <button
                    className={copy.className}
                    onClick={copy.handleCopyClick}
                    type="button"
                  >
                    {copy.title}
                  </button>
                )}
              </p>
              <p className={flowStyles.ctaLine}>
                <a className={flowStyles.ctaLink} href="https://twitter.com/">
                  Go to Twitter
                </a>
              </p>
            </div>
          </Fragment>
        )}

        {status === 'unconfirmed' && (
          <DetailedMessage
            icon="exclamation"
            heading="Attestation error:"
            message="Tweet not found"
            details="SocialKYC could not find your tweet. Please make sure you tweet the text exactly as provided. You can use the copy button to avoid any typos."
          />
        )}

        {status === 'attesting' && <AttestationProcessAnchoring />}

        {status === 'ready' && (
          <Fragment>
            <AttestationProcessReady />
            <button
              className={flowStyles.ctaButton}
              type="button"
              onClick={handleBackup}
            >
              Show credential in wallet
            </button>
          </Fragment>
        )}

        {flowError === 'unknown' && <AttestationErrorUnknown />}

        {(status === 'unconfirmed' || flowError === 'unknown') && (
          <button
            type="button"
            className={flowStyles.ctaButton}
            onClick={handleTryAgainClick}
          >
            Try again
          </button>
        )}
      </section>

      <LinkBack />
    </section>
  );
}
