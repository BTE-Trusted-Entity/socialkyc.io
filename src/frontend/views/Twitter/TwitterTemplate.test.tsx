import { describe, it, jest, expect } from '@jest/globals';

import { render } from '../../../testing/testing';
import '../../components/useCopyButton/useCopyButton.mock';

import { TwitterTemplate } from './TwitterTemplate';

jest.useFakeTimers();
jest.setSystemTime(new Date('2022-01-03T12:00:00'));

const actions = {
  handleSubmit: jest.fn(),
  handleBackup: jest.fn(),
  handleTryAgainClick: jest.fn(),
  setInputError: jest.fn(),
};

describe('TwitterTemplate', () => {
  it('should match snapshot with status=none', async () => {
    const { container } = render(
      <TwitterTemplate status="none" processing={false} {...actions} />,
    );
    expect(container).toMatchSnapshot();
  });

  it('should match snapshot with status=requested', async () => {
    const { container } = render(
      <TwitterTemplate status="requested" processing={true} {...actions} />,
    );
    expect(container).toMatchSnapshot();
  });

  it('should match snapshot with status=confirming', async () => {
    const { container } = render(
      <TwitterTemplate
        status="confirming"
        secret="0123456789"
        processing={false}
        {...actions}
      />,
    );
    expect(container).toMatchSnapshot();
  });

  it('should match snapshot with status=attesting', async () => {
    const { container } = render(
      <TwitterTemplate status="attesting" processing={false} {...actions} />,
    );
    expect(container).toMatchSnapshot();
  });

  it('should match snapshot with status=ready', async () => {
    const { container } = render(
      <TwitterTemplate status="ready" processing={false} {...actions} />,
    );
    expect(container).toMatchSnapshot();
  });

  it('should match snapshot with flowError=closed', async () => {
    const { container } = render(
      <TwitterTemplate
        status="requested"
        flowError="closed"
        processing={false}
        {...actions}
      />,
    );
    expect(container).toMatchSnapshot();
  });

  it('should match snapshot with flowError=unknown', async () => {
    const { container } = render(
      <TwitterTemplate
        status="error"
        flowError="unknown"
        processing={false}
        {...actions}
      />,
    );
    expect(container).toMatchSnapshot();
  });

  it('should match snapshot with status=unconfirmed', async () => {
    const { container } = render(
      <TwitterTemplate status="unconfirmed" processing={false} {...actions} />,
    );
    expect(container).toMatchSnapshot();
  });
});
