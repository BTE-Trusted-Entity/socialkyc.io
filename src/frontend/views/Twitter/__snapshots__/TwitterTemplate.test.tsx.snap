// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TwitterTemplate should match snapshot with flowError=closed 1`] = `
<div>
  <section
    aria-busy="false"
    class="container"
  >
    <h1
      class="heading"
    >
      Twitter Account Attestation
    </h1>
    <div
      class="explainerContainer"
    >
      <button
        aria-label="Toggle explanation"
        class="showExplainer"
        type="button"
      />
    </div>
    <section>
      <form>
        <dl
          class="profile"
        >
          <dt>
            Twitter handle:
          </dt>
          <dd>
            social_kyc_tech
          </dd>
        </dl>
        <p>
          Validity: one year (
          1/3/2023
          )
        </p>
        <section
          class="container"
        >
          <div
            class="exclamation"
          />
          <h2
            class="heading"
          >
            Signing error:
          </h2>
          <p
            class="message"
          >
            Your wallet was closed before the request was signed.
          </p>
          <p>
            Click “Continue in Wallet” to try again.
          </p>
        </section>
        <button
          class="ctaButton"
          type="submit"
        >
          Continue in wallet
        </button>
      </form>
    </section>
    <a
      class="link"
      href="/"
    >
      Back to list
    </a>
  </section>
</div>
`;

exports[`TwitterTemplate should match snapshot with flowError=unknown 1`] = `
<div>
  <section
    aria-busy="false"
    class="container"
  >
    <h1
      class="heading"
    >
      Twitter Account Attestation
    </h1>
    <div
      class="explainerContainer"
    >
      <button
        aria-label="Toggle explanation"
        class="showExplainer"
        type="button"
      />
    </div>
    <section>
      <section
        class="container"
      >
        <div
          class="exclamation"
        />
        <h2
          class="heading"
        >
          Attestation error:
        </h2>
        <p
          class="message"
        >
          Something went wrong!
        </p>
        <p>
          Click “Try Again” or reload the page or restart your browser.
        </p>
      </section>
      <button
        class="ctaButton"
        type="button"
      >
        Try again
      </button>
    </section>
    <a
      class="link"
      href="/"
    >
      Back to list
    </a>
  </section>
</div>
`;

exports[`TwitterTemplate should match snapshot with flowerror=timeout 1`] = `
<div>
  <section
    aria-busy="false"
    class="container"
  >
    <h1
      class="heading"
    >
      Twitter Account Attestation
    </h1>
    <div
      class="explainerContainer"
    >
      <button
        aria-label="Toggle explanation"
        class="showExplainer"
        type="button"
      />
    </div>
    <section>
      <section
        class="container"
      >
        <div
          class="exclamation"
        />
        <h2
          class="heading"
        >
          Attestation error:
        </h2>
        <p
          class="message"
        >
          Tweet not found
        </p>
        <p>
          SocialKYC could not find your tweet. Please make sure you tweet the text exactly as provided. You can use the copy button to avoid any typos.
        </p>
      </section>
      <button
        class="ctaButton"
        type="button"
      >
        Try again
      </button>
    </section>
    <a
      class="link"
      href="/"
    >
      Back to list
    </a>
  </section>
</div>
`;

exports[`TwitterTemplate should match snapshot with status=attesting 1`] = `
<div>
  <section
    aria-busy="false"
    class="container"
  >
    <h1
      class="heading"
    >
      Twitter Account Attestation
    </h1>
    <div
      class="explainerContainer"
    >
      <button
        aria-label="Toggle explanation"
        class="showExplainer"
        type="button"
      />
    </div>
    <section>
      <section
        class="container"
      >
        <div
          class="spinner"
        />
        <h2
          class="heading"
        >
          Attestation process:
        </h2>
        <p
          class="message"
        >
          Anchoring credential on KILT blockchain
        </p>
        <p>
          Please leave this tab open until your credential is attested.
        </p>
      </section>
    </section>
    <a
      class="link"
      href="/"
    >
      Back to list
    </a>
  </section>
</div>
`;

exports[`TwitterTemplate should match snapshot with status=authenticated 1`] = `
<div>
  <section
    aria-busy="false"
    class="container"
  >
    <h1
      class="heading"
    >
      Twitter Account Attestation
    </h1>
    <div
      class="explainerContainer"
    >
      <button
        aria-label="Toggle explanation"
        class="showExplainer"
        type="button"
      />
    </div>
    <section>
      <form>
        <dl
          class="profile"
        >
          <dt>
            Twitter handle:
          </dt>
          <dd>
            social_kyc_tech
          </dd>
        </dl>
        <p>
          Validity: one year (
          1/3/2023
          )
        </p>
        <button
          class="ctaButton"
          type="submit"
        >
          Continue in wallet
        </button>
      </form>
    </section>
    <a
      class="link"
      href="/"
    >
      Back to list
    </a>
  </section>
</div>
`;

exports[`TwitterTemplate should match snapshot with status=authenticating 1`] = `
<div>
  <section
    aria-busy="false"
    class="container"
  >
    <h1
      class="heading"
    >
      Twitter Account Attestation
    </h1>
    <div
      class="explainerContainer"
    >
      <button
        aria-label="Toggle explanation"
        class="showExplainer"
        type="button"
      />
    </div>
    <section>
      <section
        class="container"
      >
        <div
          class="spinner"
        />
        <h2
          class="heading"
        >
          Attestation process:
        </h2>
        <p
          class="message"
        >
          Starting
        </p>
        <p>
          To continue the attestation process, please Tweet the text below and then return to this page. It may take SocialKYC a moment to complete the process.
        </p>
      </section>
      <div>
        <label
          for="tweet"
        >
          Please tweet this message:
        </label>
        <p
          class="tweetContainer"
        >
          <textarea
            class="tweetInput"
            id="tweet"
            readonly=""
          >
            I use SocialKYC to represent my internet identity. @social_kyc_tech SECRET
          </textarea>
          <button
            class="copy"
            type="button"
          >
            copy to clipboard
          </button>
        </p>
        <p
          class="ctaLine"
        >
          <a
            class="ctaLink"
            href="https://twitter.com/"
            rel="noreferrer"
            target="_blank"
          >
            Go to Twitter
          </a>
        </p>
      </div>
    </section>
    <a
      class="link"
      href="/"
    >
      Back to list
    </a>
  </section>
</div>
`;

exports[`TwitterTemplate should match snapshot with status=none 1`] = `
<div>
  <section
    aria-busy="false"
    class="container"
  >
    <h1
      class="heading"
    >
      Twitter Account Attestation
    </h1>
    <div
      class="explainerContainer"
    >
      <button
        aria-label="Toggle explanation"
        class="showExplainer"
        type="button"
      />
    </div>
    <section>
      <form>
        <label>
          Your Twitter handle
          <span
            class="twitterInputContainer"
          >
            <input
              class="twitterInput"
              maxlength="100"
              name="twitterHandle"
              required=""
              type="text"
              value=""
            />
          </span>
        </label>
        <output
          class="inputError"
          hidden=""
        />
        <p>
          Validity: one year (
          1/3/2023
          )
        </p>
        <button
          class="ctaButton"
          disabled=""
          type="submit"
        >
          Continue
        </button>
      </form>
    </section>
    <a
      class="link"
      href="/"
    >
      Back to list
    </a>
  </section>
</div>
`;

exports[`TwitterTemplate should match snapshot with status=ready 1`] = `
<div>
  <section
    aria-busy="false"
    class="container"
  >
    <h1
      class="heading"
    >
      Twitter Account Attestation
    </h1>
    <div
      class="explainerContainer"
    >
      <button
        aria-label="Toggle explanation"
        class="showExplainer"
        type="button"
      />
    </div>
    <section>
      <section
        class="container"
      >
        <div
          class="checkmark"
        />
        <h2
          class="heading"
        >
          Attestation process:
        </h2>
        <p
          class="message"
        >
          Credential is ready
        </p>
        <p>
          We recommend that you back up your credential now.
        </p>
      </section>
      <button
        class="ctaButton"
        type="button"
      >
        Show credential in wallet
      </button>
    </section>
    <a
      class="link"
      href="/"
    >
      Back to list
    </a>
  </section>
</div>
`;
